clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java

```diff
@@ -314,64 +314,73 @@ public void testFormatConversionWithPartialMessage() throws IOException {
+        builder.close();
 
-        // down conversion for non-compressed messages
-        try (FileRecords fileRecords = FileRecords.open(tempFile())) {
-            fileRecords.append(builder.build());
-            fileRecords.flush();
-            Records convertedRecords = fileRecords.downConvert(RecordBatch.MAGIC_VALUE_V0);
-            verifyConvertedMessageSet(records, offsets, convertedRecords, RecordBatch.MAGIC_VALUE_V0);
-        }
-    }
-
-    @Test
-    public void testConvertCompressedToMagic0() throws IOException {
-        List<Long> offsets = Arrays.asList(0L, 2L);
-        List<SimpleRecord> records = Arrays.asList(
-                new SimpleRecord(1L, "k1".getBytes(), "hello".getBytes()),
-                new SimpleRecord(2L, "k2".getBytes(), "goodbye".getBytes()));
+        builder = MemoryRecords.builder(buffer, RecordBatch.MAGIC_VALUE_V1, compressionType,
+                TimestampType.CREATE_TIME, 0L);
+        for (int i = 2; i < 4; i++)
+            builder.appendWithOffset(offsets.get(i), records.get(i));
+        builder.close();
 
-        ByteBuffer buffer = ByteBuffer.allocate(512);
-        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, RecordBatch.MAGIC_VALUE_V1, CompressionType.GZIP,
+        builder = MemoryRecords.builder(buffer, RecordBatch.MAGIC_VALUE_V2, compressionType,
                 TimestampType.CREATE_TIME, 0L);
-        for (int i = 0; i < offsets.size(); i++)
+        for (int i = 4; i < 6; i++)
             builder.appendWithOffset(offsets.get(i), records.get(i));
+        builder.close();
+
+        buffer.flip();
 
-        // down conversion for compressed messages
+        // down conversion for non-compressed messages
         try (FileRecords fileRecords = FileRecords.open(tempFile())) {
-            fileRecords.append(builder.build());
+            fileRecords.append(MemoryRecords.readableRecords(buffer));
             fileRecords.flush();
-            Records convertedRecords = fileRecords.downConvert(RecordBatch.MAGIC_VALUE_V0);
-            verifyConvertedMessageSet(records, offsets, convertedRecords, RecordBatch.MAGIC_VALUE_V0);
+            Records convertedRecords = fileRecords.downConvert(toMagic);
+            verifyConvertedMessageSet(records, offsets, convertedRecords, compressionType, toMagic);
         }
     }
 
     private void verifyConvertedMessageSet(List<SimpleRecord> initialRecords,
                                            List<Long> initialOffsets,
                                            Records convertedRecords,
+                                           CompressionType compressionType,
                                            byte magicByte) {
         int i = 0;
         for (RecordBatch batch : convertedRecords.batches()) {
-            assertEquals("magic byte should be " + magicByte, magicByte, batch.magic());
+            assertTrue("Magic byte should be lower than or equal to " + magicByte, batch.magic() <= magicByte);
```

Comment:
We only convert messages when necessary. So if we down-convert a batch to version 1, and we have messages which are magic 0, we won't up-convert them to 1.